pkg/issuer/acme/setup.go

```diff
@@ -135,21 +161,19 @@ func (a *Acme) Setup(ctx context.Context) error {
 	rawServerURL := a.issuer.GetSpec().ACME.Server
 	parsedServerURL, err := url.Parse(rawServerURL)
 	if err != nil {
-		r := "InvalidURL"
-		s := fmt.Sprintf("Failed to parse existing ACME server URI %q: %v", rawServerURL, err)
-		a.recorder.Eventf(a.issuer, corev1.EventTypeWarning, r, s)
-		apiutil.SetIssuerCondition(a.issuer, a.issuer.GetGeneration(), v1.IssuerConditionReady, cmmeta.ConditionFalse, r, s)
+		reason = errorInvalidURL
+		msg = fmt.Sprintf(messageTemplateFailedToParseURL, rawServerURL, err)
+		a.recorder.Event(a.issuer, corev1.EventTypeWarning, errorInvalidURL, msg)
```

Comment:
I don't understand why it is only from this point in the function that we generate Kubernetes warning events to accompany the False Ready condition.
Why not create Events for the each of the error cases in the switch statement above?

Reply:
I see that we already throw an error in the caller every time `Setup` returns with an error [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/issuers/sync.go#L61).

I would be inclined to do it either just in caller or just in `Setup`. Or perhaps generate success events in `Setup` (as we don't pass any info to caller in that scenario), but leave the error events to be generated by caller.

Reply:
Ah, thanks for explaining. So we record a warning event in this case, because we *are not* returning an error for this event and therefore the  caller will not generate record the warning event.